From 3540bf8a5bfddb84851795cac50ee8b347709866 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 8 Mar 2022 23:53:48 +0100
Subject: [PATCH 15/16] user32: Send WM_*FOCUS messages even if already
 focused.

---
 dlls/win32u/input.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

#diff --git a/dlls/user32/tests/msg.c b/dlls/user32/tests/msg.c
#index 11111111111..11111111111 100644
#--- a/dlls/user32/tests/msg.c
#+++ b/dlls/user32/tests/msg.c
#@@ -15220,15 +15220,15 @@ static void test_interthread_set_foreground_window(struct test_set_foreground_wi
#         { hwnd2, TRUE, FALSE, FALSE,   FALSE, WmEmptySeq, FALSE, sequence_3,  TRUE, sequence_0,  TRUE, hwnd1 },
#         { hwnd2, FALSE, TRUE, FALSE,   FALSE, sequence_3, FALSE, WmEmptySeq,  TRUE, sequence_0,  TRUE, hwnd1 },
#         { hwnd2, TRUE, TRUE, FALSE,    FALSE, sequence_4, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
#-        { hwnd2, FALSE, FALSE, TRUE,    TRUE, sequence_5, FALSE, WmEmptySeq,  TRUE, sequence_0,  TRUE, hwnd1 },
#-        { hwnd2, TRUE, FALSE, TRUE,     TRUE, sequence_6, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
#+        { hwnd2, FALSE, FALSE, TRUE,   FALSE, sequence_5, FALSE, WmEmptySeq,  TRUE, sequence_0,  TRUE, hwnd1 },
#+        { hwnd2, TRUE, FALSE, TRUE,    FALSE, sequence_6, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
# 
#         { hwnd0, FALSE, FALSE, FALSE,  FALSE, WmEmptySeq, FALSE, sequence_3, FALSE, sequence_0, FALSE, hwnd1 },
#         { hwnd0, TRUE, FALSE, FALSE,   FALSE, WmEmptySeq, FALSE, sequence_3,  TRUE, sequence_0,  TRUE, hwnd1 },
#         { hwnd0, FALSE, TRUE, FALSE,   FALSE, sequence_3, FALSE, WmEmptySeq, FALSE, sequence_0, FALSE, hwnd1 },
#         { hwnd0, TRUE, TRUE, FALSE,    FALSE, sequence_4, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
#-        { hwnd0, FALSE, FALSE, TRUE,    TRUE, sequence_5, FALSE, WmEmptySeq, FALSE, sequence_0, FALSE, hwnd1 },
#-        { hwnd0, TRUE, FALSE, TRUE,     TRUE, sequence_6, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
#+        { hwnd0, FALSE, FALSE, TRUE,   FALSE, sequence_5, FALSE, WmEmptySeq, FALSE, sequence_0, FALSE, hwnd1 },
#+        { hwnd0, TRUE, FALSE, TRUE,    FALSE, sequence_6, FALSE, sequence_1, FALSE, sequence_2, FALSE, hwnd0 },
#     };
# 
#     DWORD i, res;
diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -1769,7 +1769,7 @@ BOOL WINAPI release_capture(void)
  *
  * Change the focus window, sending the WM_SETFOCUS and WM_KILLFOCUS messages
  */
-static HWND set_focus_window( HWND hwnd )
+static HWND set_focus_window( HWND hwnd, HWND last_focus )
 {
     HWND previous = 0, ime_hwnd;
     BOOL ret;
@@ -1782,7 +1782,8 @@ static HWND set_focus_window( HWND hwnd )
     }
     SERVER_END_REQ;
     if (!ret) return 0;
-    if (previous == hwnd) return previous;
+    if (!last_focus) last_focus = previous;
+    if (last_focus == hwnd) return previous;
 
     if (previous)
     {
@@ -1926,7 +1927,7 @@ static BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus )
         {
             /* this line exists to keep this patch from applying in the wrong place */
             if (!info.hwndFocus || !hwnd || NtUserGetAncestor( info.hwndFocus, GA_ROOT ) != hwnd)
-                set_focus_window( hwnd );
+                set_focus_window( hwnd, 0 );
         }
     }
 
@@ -2020,7 +2021,7 @@ HWND WINAPI NtUserSetFocus( HWND hwnd )
     }
 
     /* change focus and send messages */
-    return set_focus_window( hwnd );
+    return set_focus_window( hwnd, previous );
 }
 
 /*******************************************************************
-- 
2.40.0

